File: packages/shared/src/athlete.ts
import type { BeachPosition, CourtPosition, Modality, Sex } from "./enums";

export interface AthleteAttributes {
  serve: number;
  attack: number;
  block: number;
  defense: number;
  reception: number;
  setting: number;
  speed: number;
  jump: number;
  stamina: number;
  positioning: number;
  decision: number;
  concentration: number;
  competitiveness: number;
}

export interface Athlete {
  id: string;
  club_id: string | null;
  first_name: string;
  last_name: string;
  country: string;
  city: string | null;
  birth_date: string;
  height_cm: number;
  weight_kg: number;
  handedness: "left" | "right";
  sex: Sex;
  modality: Modality;
  court_position: CourtPosition | null;
  beach_position: BeachPosition | null;
  current_ability: number;
  potential_ability: number;
  morale: number;
  fatigue: number;
  form: number;
  market_value: number;
  sale_value: number;
  wins: number;
  losses: number;
  is_custom: boolean;
  is_injured: boolean;
  level: number;
  level_xp: number;
  condition: AthleteCondition;
  rest_games_left: number;
  injured_until: string | null;
  last_trained_on: string | null;
  for_sale: boolean;
  expires_at: string | null;
  attributes: AthleteAttributes | null;
}

export type AthleteCondition = "ok" | "fatigued" | "injured";

export const CONDITION_LABEL: Record<AthleteCondition, string> = {
  ok: "Pronto",
  fatigued: "Fadigado",
  injured: "Lesionado",
};

/** Treinos disponíveis (1 por dia por atleta). Espelha app/engine/training.py. */
export const TRAININGS: { key: string; label: string; hint: string }[] = [
  { key: "saque", label: "Saque", hint: "+Saque, −Recepção" },
  { key: "recepcao", label: "Recepção", hint: "+Recepção, −Saque" },
  { key: "passe", label: "Passe", hint: "+Levantamento/Posicionamento, −Bloqueio" },
  { key: "levantamento", label: "Levantamento", hint: "+Levantamento, −Resistência" },
  { key: "manchete", label: "Manchete", hint: "+Defesa, −Ataque" },
  { key: "ataque", label: "Ataque", hint: "+Ataque, −Defesa" },
  { key: "cortada", label: "Cortada", hint: "+Ataque/Impulsão, −Recepção" },
  { key: "impulsao", label: "Impulsão", hint: "+Impulsão, −Resistência" },
  { key: "deslocamento", label: "Deslocamento", hint: "+Velocidade, −Impulsão" },
  { key: "defesa", label: "Defesa", hint: "+Defesa/Posicionamento, −Saque" },
];

export interface Club {
  id: string;
  owner_id: string | null;
  name: string;
  short_name: string | null;
  crest_url: string | null;
  country: string;
  city: string | null;
  modality: Modality;
  reputation: number;
  fanbase: number;
  is_cpu: boolean;
}

export const POSITION_LABEL: Record<string, string> = {
  // quadra
  setter: "Levantador",
  opposite: "Oposto",
  outside: "Ponteiro",
  middle: "Central",
  libero: "Líbero",
  // praia
  defender: "Defensor",
  blocker: "Bloqueador",
  universal: "Universal",
};

export const ATTRIBUTE_LABEL: Record<keyof AthleteAttributes, string> = {
  serve: "Saque",
  attack: "Ataque",
  block: "Bloqueio",
  defense: "Defesa",
  reception: "Recepção",
  setting: "Levantamento",
  speed: "Velocidade",
  jump: "Impulsão",
  stamina: "Resistência",
  positioning: "Posicionamento",
  decision: "Tomada de decisão",
  concentration: "Concentração",
  competitiveness: "Mentalidade",
};

File: apps/api/app/schemas/athlete.py
"""Schemas Pydantic (DTOs) para atletas."""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

from app.enums import BeachPosition, CourtPosition, Handedness, Modality, Sex


class AttributesOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    serve: int
    attack: int
    block: int
    defense: int
    reception: int
    setting: int
    speed: int
    jump: int
    stamina: int
    positioning: int
    decision: int
    concentration: int
    competitiveness: int


class AthleteOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    club_id: uuid.UUID | None
    first_name: str
    last_name: str
    country: str
    city: str | None
    birth_date: date
    height_cm: int
    weight_kg: int
    handedness: Handedness
    sex: Sex
    modality: Modality
    court_position: CourtPosition | None
    beach_position: BeachPosition | None
    current_ability: int
    potential_ability: int
    morale: int
    fatigue: int
    form: int
    market_value: int
    sale_value: int
    wins: int
    losses: int
    is_custom: bool
    is_injured: bool
    # Progressão e condição física.
    level: int = 1
    level_xp: int = 0
    condition: str = "ok"            # "ok" | "fatigued" | "injured"
    rest_games_left: int = 0
    injured_until: datetime | None = None
    last_trained_on: date | None = None
    for_sale: bool = False
    expires_at: datetime | None = None
    listing_id: uuid.UUID | None = None  # veio de um anúncio (não vende, não treina)
    attributes: AttributesOut | None = None


class AthleteCreate(BaseModel):
    first_name: str
    last_name: str
    country: str = "BRA"
    city: str | None = None
    birth_date: date
    height_cm: int = Field(ge=140, le=230)
    weight_kg: int = Field(ge=40, le=160)
    handedness: Handedness = Handedness.RIGHT
    modality: Modality
    court_position: CourtPosition | None = None
    beach_position: BeachPosition | None = None
    club_id: uuid.UUID | None = None


class GenerateRequest(BaseModel):
    modality: Modality
    count: int = Field(default=10, ge=1, le=200)
    country: str | None = None
    seed: int | None = None
    club_id: uuid.UUID | None = None  # se nulo, gera agentes livres


class AttributesIn(BaseModel):
    serve: int = Field(default=50, ge=1, le=99)
    attack: int = Field(default=50, ge=1, le=99)
    block: int = Field(default=50, ge=1, le=99)
    defense: int = Field(default=50, ge=1, le=99)
    reception: int = Field(default=50, ge=1, le=99)
    setting: int = Field(default=50, ge=1, le=99)
    speed: int = Field(default=50, ge=1, le=99)
    jump: int = Field(default=50, ge=1, le=99)
    stamina: int = Field(default=50, ge=1, le=99)
    positioning: int = Field(default=50, ge=1, le=99)
    decision: int = Field(default=50, ge=1, le=99)
    concentration: int = Field(default=50, ge=1, le=99)
    competitiveness: int = Field(default=50, ge=1, le=99)


class CustomAthleteCreate(BaseModel):
    """Atleta personalizado criado pelo jogador (Mercado → Contratações)."""

    first_name: str = Field(min_length=1, max_length=40)
    last_name: str = Field(min_length=1, max_length=40)
    country: str = "BRA"
    sex: Sex
    modality: Modality
    court_position: CourtPosition | None = None
    beach_position: BeachPosition | None = None
    height_cm: int = Field(default=190, ge=140, le=230)
    weight_kg: int = Field(default=85, ge=40, le=160)
    birth_date: date | None = None
    attributes: AttributesIn = Field(default_factory=AttributesIn)
